File: modules/openlmis-web/src/main/java/org/openlmis/web/controller/SupervisoryNodeController.java
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

package org.openlmis.web.controller;

import lombok.NoArgsConstructor;
import org.openlmis.core.domain.Pagination;
import org.openlmis.core.domain.SupervisoryNode;
import org.openlmis.core.exception.DataException;
import org.openlmis.core.service.SupervisoryNodeService;
import org.openlmis.core.web.OpenLmisResponse;
import org.openlmis.core.web.controller.BaseController;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;

import javax.servlet.http.HttpServletRequest;
import java.util.List;

import static org.springframework.http.HttpStatus.BAD_REQUEST;
import static org.springframework.web.bind.annotation.RequestMethod.*;

/**
 * This controller handle endpoints to list, search, create, update supervisory nodes.
 */

@Controller
@NoArgsConstructor
public class SupervisoryNodeController extends BaseController {

  public static final String SUPERVISORY_NODES = "supervisoryNodes";

  @Autowired
  private SupervisoryNodeService supervisoryNodeService;

  @RequestMapping(value = "/supervisory-nodes", method = GET)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_USER')")
  public ResponseEntity<OpenLmisResponse> getAll() {
    return OpenLmisResponse.response(SUPERVISORY_NODES, supervisoryNodeService.getAll());
  }

  @RequestMapping(value = "/supervisory-nodes/list", method = GET)
  public ResponseEntity<OpenLmisResponse> getAllReadOnly() {
    return OpenLmisResponse.response(SUPERVISORY_NODES, supervisoryNodeService.getAll());
  }

  @RequestMapping(value = "/paged-search-supervisory-nodes", method = GET)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_SUPERVISORY_NODE')")
  public ResponseEntity<OpenLmisResponse> searchSupervisoryNode(@RequestParam(value = "page", required = true, defaultValue = "1") Integer page,
                                                                @RequestParam(required = true) String param,
                                                                @RequestParam(required = true) Boolean parent) {
    ResponseEntity<OpenLmisResponse> response = OpenLmisResponse.response(SUPERVISORY_NODES,
      supervisoryNodeService.getSupervisoryNodesBy(page, param, parent));
    Pagination pagination = supervisoryNodeService.getPagination(page);
    pagination.setTotalRecords(supervisoryNodeService.getTotalSearchResultCount(param, parent));
    response.getBody().addData("pagination", pagination);
    return response;
  }

  @RequestMapping(value = "/supervisory-nodes/{id}", method = GET)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_SUPERVISORY_NODE')")
  public SupervisoryNode getById(@PathVariable(value = "id") Long id) {
    return supervisoryNodeService.getSupervisoryNode(id);
  }

  @RequestMapping(value = "/search-supervisory-nodes", method = GET, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_SUPERVISORY_NODE, MANAGE_REQUISITION_GROUP')")
  public List<SupervisoryNode> getFilteredNodes(@RequestParam(value = "searchParam") String param) {
    return supervisoryNodeService.getFilteredSupervisoryNodesByName(param);
  }

  @RequestMapping(value = "/supervisory-nodes", method = POST)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_SUPERVISORY_NODE')")
  public ResponseEntity<OpenLmisResponse> insert(@RequestBody SupervisoryNode supervisoryNode,
                                                 HttpServletRequest request) {
    ResponseEntity<OpenLmisResponse> response;
    Long userId = loggedInUserId(request);
    supervisoryNode.setCreatedBy(userId);
    supervisoryNode.setModifiedBy(userId);
    try {
      supervisoryNodeService.save(supervisoryNode);
    } catch (DataException de) {
      response = OpenLmisResponse.error(de, BAD_REQUEST);
      return response;
    }
    response = OpenLmisResponse.success(
      messageService.message("message.supervisory.node.created.success", supervisoryNode.getName()));
    response.getBody().addData("supervisoryNodeId", supervisoryNode.getId());
    return response;
  }

  @RequestMapping(value = "/supervisory-nodes/{id}", method = PUT, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_SUPERVISORY_NODE')")
  public ResponseEntity<OpenLmisResponse> update(@RequestBody SupervisoryNode supervisoryNode,
                                                 @PathVariable(value = "id") Long supervisoryNodeId,
                                                 HttpServletRequest request) {
    ResponseEntity<OpenLmisResponse> response;
    Long userId = loggedInUserId(request);
    supervisoryNode.setModifiedBy(userId);
    supervisoryNode.setId(supervisoryNodeId);
    try {
      supervisoryNodeService.save(supervisoryNode);
    } catch (DataException de) {
      response = OpenLmisResponse.error(de, BAD_REQUEST);
      return response;
    }
    response = OpenLmisResponse.success(
      messageService.message("message.supervisory.node.updated.success", supervisoryNode.getName()));
    response.getBody().addData("supervisoryNodeId", supervisoryNode.getId());
    return response;
  }

  @RequestMapping(value = "/topLevelSupervisoryNodes", method = GET, headers = ACCEPT_JSON)
  @PreAuthorize("@permissionEvaluator.hasPermission(principal,'MANAGE_SUPPLY_LINE')")
  public List<SupervisoryNode> searchTopLevelSupervisoryNodesByName(@RequestParam(value = "searchParam") String param) {
    return supervisoryNodeService.searchTopLevelSupervisoryNodesByName(param);
  }
}

File: modules/openlmis-web/src/main/webapp/public/js/supervisory-node/controller/supervisory-node-controller.js
/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2013 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU Affero General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more details.
 * You should have received a copy of the GNU Affero General Public License along with this program.  If not, see http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org.
 */

function SupervisoryNodeController($scope, $location, supervisoryNode, ParentSupervisoryNodes, SupervisoryNodes) {
  $scope.supervisoryNode = supervisoryNode || {};
  $scope.$parent.message = "";

  var reset = function () {
    $scope.query = undefined;
    $scope.nodeResultCount = undefined;
    $scope.filteredFacilities = undefined;
    $scope.previousQuery = undefined;
  };

  var compareQuery = function () {
    if (!isUndefined($scope.previousQuery)) {
      return $scope.query.substr(0, 3) !== $scope.previousQuery.substr(0, 3);
    }
    return true;
  };

  $scope.toggleSlider = function(){
    $scope.showSlider = !$scope.showSlider;
    $scope.extraParams = {"virtualFacility": null, "enabled": true};
  };

  $scope.associate = function (facility) {
    $scope.supervisoryNode.facility = facility;
    $scope.showSlider = !$scope.showSlider;
  };

  $scope.showParentNodeSearchResults = function () {
    if ($scope.query === undefined || $scope.query.length < 3) return;

    if (compareQuery()) {
      ParentSupervisoryNodes.get({"searchParam": $scope.query}, function (data) {
        $scope.parentNodes = _.reject(data.supervisoryNodeList, function (node) {
          return $scope.supervisoryNode === undefined ? false : ($scope.supervisoryNode.id && node.code == $scope.supervisoryNode.code);
        });
        $scope.filteredNodes = $scope.parentNodes;
        $scope.previousQuery = $scope.query;
        $scope.nodeResultCount = $scope.filteredNodes.length;
      });
    }
    else {
      $scope.filteredNodes = _.filter($scope.parentNodes, function (node) {
        return node.name.toLowerCase().indexOf($scope.query.toLowerCase()) !== -1;
      });
      $scope.nodeResultCount = $scope.filteredNodes.length;
    }
  };

  $scope.setSelectedParentNode = function (node) {
    $scope.supervisoryNode.parent = node;
    reset();
  };

  $scope.deleteParentNode = function () {
    $scope.supervisoryNode.parent = undefined;
    angular.element("#searchParentNode").focus();
  };

  $scope.cancel = function () {
    $scope.$parent.message = "";
    $scope.$parent.supervisoryNodeId = undefined;
    $location.path('#/search');
  };

  var success = function (data) {
    $scope.error = "";
    $scope.$parent.message = data.success;
    $scope.$parent.supervisoryNodeId = data.supervisoryNodeId;
    $scope.showError = false;
    $location.path('');
  };

  var error = function (data) {
    $scope.$parent.message = "";
    $scope.error = data.data.error;
    $scope.showError = true;
  };

  $scope.save = function () {
    if ($scope.supervisoryNodeForm.$error.required || !$scope.supervisoryNode.facility) {
      $scope.showError = true;
      $scope.error = "form.error";
      return;
    }
    if ($scope.supervisoryNode.id) {
      SupervisoryNodes.update({id: $scope.supervisoryNode.id}, $scope.supervisoryNode, success, error);
    }
    else {
      SupervisoryNodes.save({}, $scope.supervisoryNode, success, error);
    }
  };
}

SupervisoryNodeController.resolve = {
  supervisoryNode: function ($q, $route, $timeout, SupervisoryNodes) {
    if ($route.current.params.id === undefined) return undefined;

    var deferred = $q.defer();
    var supervisoryNodeId = $route.current.params.id;

    $timeout(function () {
      SupervisoryNodes.get({id: supervisoryNodeId}, function (data) {
        deferred.resolve(data.supervisoryNode);
      }, {});
    }, 100);
    return deferred.promise;
  }
};
